lib/class-wp-rest-pattern-directory-controller.php

```diff
@@ -0,0 +1,274 @@
+								'required'    => true,
+								'minLength'   => 1,
+							),
+						)
+					),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Checks whether a given request has permission to install and activate plugins.
+	 *
+	 * @since 5.7.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|bool True if the request has permission, WP_Error object otherwise.
+	 */
+	public function get_items_permissions_check( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedParameter
+		if ( current_user_can( 'edit_posts' ) ) {
+			return true;
+		}
+
+		return new WP_Error(
+			'rest_pattern_directory_cannot_view',
+			__( 'Sorry, you are not allowed to browse the local block pattern directory. Please use https://api.wordpress.org/patterns/1.0/ for publicly-available pattern data.', 'gutenberg' ),
+			array( 'status' => rest_authorization_required_code() )
+		);
+	}
+
+	/**
+	 * Search and retrieve block patterns metadata
+	 *
+	 * @since 5.7.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) {
+		switch ( $request->get_route() ) {
+			default:
+			case '/wp/v2/pattern-directory':
+				$category   = (int) $request->get_param( 'category' );
+				$query_args = array(
+					'action' => 'get_patterns',
+				);
+
+				if ( $category ) {
+					$query_args['pattern-categories'] = $category;
+				}
+
+				break;
+
+			case '/wp/v2/pattern-directory/search':
+				$query_args = array(
+					'action' => 'query_patterns',
+					'search' => sanitize_text_field( $request['term'] ),
+				);
+
+				break;
+		}
+
+		$api_url = add_query_arg(
+			$query_args,
+			'http://api.wordpress.org/patterns/1.0/'
+		);
+
+		if ( wp_http_supports( array( 'ssl' ) ) ) {
+			$api_url = set_url_scheme( $api_url, 'https' );
+		}
+
+		$response     = wp_remote_get( $api_url );
+		$raw_patterns = json_decode( wp_remote_retrieve_body( $response ) );
+
+		if ( is_wp_error( $response ) ) {
+			$response->add_data( array( 'status' => 500 ) );
+
+			return $response;
+		}
+
+		$prepared_patterns = array();
+
+		if ( $raw_patterns ) {
+			foreach ( $raw_patterns as $pattern ) {
+				$prepared_patterns[] = $this->prepare_item_for_response( $pattern, $query_args['action'] );
+			}
+		}
+
+		return rest_ensure_response( $prepared_patterns );
+	}
+
+	/**
+	 * Parse pattern metadata for a pattern, and prepare it for an API response.
+	 *
+	 * @since 5.7.0
+	 *
+	 * @param array  $raw_pattern A pattern from api.wordpress.org, before any changes.
+	 * @param string $action      The type of request that resulted in the raw patterns. Either 'get_patterns' or 'query_patterns'.
+	 *
+	 * @return array
+	 */
+	public function prepare_item_for_response( $raw_pattern, $action ) {
+		switch ( $action ) {
+			default:
```

Comment:
This is a very much a nit, but in most `switch` statements, the `default` is the last case. Could we maybe reverse these so `query_patterns` is first?